02-microarray/pathway-analysis_microarray_03_gsea.Rmd

```diff
@@ -0,0 +1,314 @@
+
+The annotation package `org.Dr.eg.db` contains information for different identifiers [@@Carlson2019-zebrafish].
+`org.Dr.eg.db` is specific to _Danio rerio_ -- this is what the `Dr` in the package name is referencing.
+
+We can see what types of IDs are available to us in an annotation package with `keytypes()`.
+
+```{r}
+keytypes(org.Dr.eg.db)
+```
+
+Even though we'll use this package to convert from Ensembl gene IDs (`ENSEMBL`) to Entrez IDs (`ENTREZID`), we could just as easily use it to convert from an Ensembl transcript ID (`ENSEMBLTRANS`) to gene symbols (`SYMBOL`).
+
+Take a look at our other gene identifier conversion examples for examples with different species and gene ID types:
+[the microarray example](https://alexslemonade.github.io/refinebio-examples/02-microarray/gene-id-annotation_microarray_01_ensembl.html) and [the RNA-seq example](https://alexslemonade.github.io/refinebio-examples/03-rnaseq/gene-id-annotation_rnaseq_01_ensembl.html).
+
+The function we will use to map from Ensembl gene IDs to Entrez gene IDs is called `mapIds()`.
+
+Let's create a column in the `dge_df` data frame that shows the mapped Entrez IDs along with the information for the respective Ensembl IDs
+
+```{r}
+# We would like a data frame we can join to the differential expression stats
+dge_df <- dge_df %>%
+  # Create a new column 'entrez_id' that contains the Entrez IDs returned by
+  # mapIds()
+  dplyr::mutate(entrez_id = mapIds(org.Dr.eg.db, # Replace with annotation package for the organism relevant to your data
+    keys = dge_df$Gene,
+    column = "ENTREZID", # Replace with the type of gene identifiers you would like to map to
+    keytype = "ENSEMBL", # Replace with the type of gene identifiers in your data
+    multiVals = "first" # Where there are multiple mappings, the first mapped value will be returned
+  )) %>%
+  # If an Ensembl gene identifier doesn't map to a Entrez gene identifier, drop that
+  # from the data frame
+  dplyr::filter(!is.na(entrez_id))
+```
+
+This message is letting us know that sometimes Ensembl gene identifiers will map to multiple Entrez IDs.
+In this case, it's also possible that a Entrez ID will map to multiple Ensembl IDs.
+
+Let's see a preview of the new `dge_df`. 
+
+```{r}
+head(dge_df)
+```
+
+Now let's look at how many duplicated Entrez IDs we have, if any.
+
+```{r}
+dge_df %>%
+  # Group by Entrez IDs
+  dplyr::group_by(entrez_id) %>%
+  # Count the number of rows in the data frame for each Entrez ID
+  dplyr::count() %>%
+  # Filter out the IDs with multiple mappings
+  dplyr::filter(n > 1)
+```
+
+It looks like we have two Entrez IDs that were mapped to multiple Ensembl IDs.
+For the purpose of performing GSEA later in this notebook, we will filter out the duplicated Entrez IDs.
+For more about how to explore this, take a look at our [microarray gene ID conversion example](https://alexslemonade.github.io/refinebio-examples/02-microarray/gene-id-annotation_microarray_01_ensembl.html).
```

Comment:
I wonder if we want to spend as much time on the gene ID mapping since we have separate examples for that? We could just skip right to using `multiVals = "filter"` since this is essentially what you are doing here but manually. 

That being said, we may not want to drop data here and I don't think we are as particular about which gene ID is used, so maybe we should just simplify,  use `multiVals = "first"`, tell them to see the mapping examples for more info and then move on?
https://www.rdocumentation.org/packages/AnnotationDbi/versions/1.30.1/topics/AnnotationDb-objects

Reply:
I believe I implemented `multiVals = "first"` as you suggest here in the last commit. I used an `inner_join()` to join the expression data as to not have NAs in the `entrez_id` column (which may pose an issue later when running `GSEA()`). Please let me know if this is what you intended @cansavvy or if I should make any further changes here.

Reply:
I will note however, that this still leaves us with two duplicated Entrez IDs that map to multiple Ensembl IDs resulting in the following warning message when running `GSEA()` later in the notebook: `There are duplicate gene names, fgsea may produce unexpected results.`

Reply:
We should probably show users that these duplicates exist then and how to deal with them. The fact that there are only two instances of this is kinda of annoying but its probably good we have this come up so we can show users how to deal with it -- in other datasets or species its possible it will come up more (or maybe not at all). 

I think we should incorporate two steps (that should maybe be their own chunk). 

1) Show users how to test for if there are multiple entrez ids. A TRUE/FALSE like `any(duplicated())` would probably work. 

2) Show one way that you can decide on which entrez id's data to keep (here's where there could be a lot of ways to do this but we will just have to pick one that we think will be generally useful in most contexts. 
I think an okay way to do this would be to keep the data for the entrez gene id with the higher t value (or lower p value) since it will be of greater interest. 

May be good to get a @jashapiro opinion on this.

Reply:
> I think an okay way to do this would be to keep the data for the entrez gene id with the higher t value (or lower p value)

Picking whatever entry has the larger absolute value for the stat we use for ranking makes sense to me. (Are we still using LFC?) A caveat we should point out is that the genes that have duplicate identifiers could be enriched in a particular pathway/gene set and you may get an overly optimistic view of how perturbed that pathway is using this approach.

Reply:
>  Are we still using LFC?

There is log fold change. And based on the [draft PR](https://github.com/AlexsLemonade/refinebio-examples/pull/339/files) LFC will be used for GSEA.

Reply:
We are using the `t` value now (per your [comment on the draft PR](https://github.com/AlexsLemonade/refinebio-examples/pull/339/files#r513817275) @jaclyn-taroni).

In the [last commit](https://github.com/AlexsLemonade/refinebio-examples/pull/345/commits/6fa5e4319a5707199f22036d69a656198bd083a9), I added two steps, one to check for duplicate identifiers and the other to sort by `t` and remove the lower duplicate value.

Let me know if you think this is the best approach given the suggestions above @cansavvy and @jaclyn-taroni.

Reply:
Did you find any evidence to support my thought that _t_ would be more "standard?" That was mostly based on recollection.

Reply:
Linking a comment that answers this question from PR 2 of 2 #347 here:

"""
The decision to create the pre-ranked gene vector based on the t-statistic rather than log fold change was explored based on [this comment from the draft PR](https://github.com/AlexsLemonade/refinebio-examples/pull/339#discussion_r513817275) and eventually made based on what is recommended in [Discovering statistically significant pathways in expression profiling studies](https://www.pnas.org/content/pnas/102/38/13544.full.pdf) and the explanation from a [biostars forum](https://www.biostars.org/p/200715/) which says "Try to understand how they relate to the question you're interested in e.g. if you're most interested in effect size then the fold change is what you should use but if you're more interested in statistical significance then look for one of the statistics taking into consideration the assumptions they make e.g. t-test" as I believe we are encouraging users to look at statistical significance here.
"""